### DISCUSSION TITLE
Undo / Redo in route planning

### DISCUSSION THREAD

**Discussion by 0709wiwiwi** [devemux86/cruiser#199] [Q&A] [answered]
Undo Redo on map button replace the unhandy history button.

Provides at the routeplanning an undo redo button action.
The immediate visible on map interaction is fast and useful.

**Comment by devemux86** [comment]
@0709wiwiwi thanks for the idea!

It is already in my implementation list.

**Comment by devemux86** [comment]
Undo / Redo in route planning will be available in the next Cruiser and it works **offline**.

The undo / redo buttons are in the routing dialog, so that the screen is not filled with extra buttons.
Do you think this is a good place for them?

With undo / redo available, do we still need the routing history screen?

(Desktop Cruiser will have the undo / redo in the right click menu)

<img src="https://github.com/devemux86/cruiser/assets/3484020/a12df1d0-6e9b-431d-9d9d-f09c4a4668a7" width="50%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/e306815e-756e-4b2b-bc9b-3a88916e36d0" width="50%"/>

**Reply by 0709wiwiwi** [reply]
The Undo redo buttons only have to appear per active planning session.

**Reply by devemux86** [reply]
> The Undo redo buttons only have to appear per active planning session.

What is an "active" planning session? What makes it turn on and off?

This is not so easy to design properly. :slightly_smiling_face: 
Apps should have full screen maps, no half screens and no fixed buttons.

e.g. we made a route change, then we want a clean map for orientation, ...
How is this possible? We do not want a permanent undo button all the time.

**Reply by 0709wiwiwi** [reply]
Example [Undo Redo buttons](https://youtube.com/shorts/qYJQPUBYazk). 

Undo Redo buttons browse the history list in map view.

![image](https://github.com/devemux86/cruiser/assets/54167201/acf9629d-a0ce-40bc-b815-7e1720aaba77)

**Reply by devemux86** [reply]
@0709wiwiwi please see my question above:

> e.g. we made a route change, then we want a clean map for orientation, ...
How is this possible? We do not want a permanent undo button all the time.

We make route changes, there is history, but we stop the route planning.
We do not want the undo button anymore, we want the full screen map.
What can be done?

Also the import - when a route exists - is making history > route changes.

**Reply by devemux86** [reply]
Like the waypoint crosshair + position buttons,
undo / redo should only be visible when they are called by the user, when they are needed.

The above implementation does this.
We can see a different or better implementation, but it should solve workflows as asked above. :slightly_smiling_face:

**Reply by devemux86** [reply]
@0709wiwiwi indeed nothing should be considered final with Cruiser!

Everything can change based on your reports and our discussions. :slightly_smiling_face:
I always need everyone's feedback so we can improve the experience.

**Reply by MotoUKRider** [reply]
This is interesting - I know I deleted a Waypoint in the list using the old buttons before the "swipe" came along, and longed for an "undo idiotic last action" button :)

The on-map idea looks similar to how the Waypoints are offered, cycling through them.  Perhaps a small on-screen label (between the 2 buttons) like: "1/10", "2/10" etc to keep track of how far back you're going?  I'm just trying to think of the minimum "taps" required to cycle through those historic edits before being able to see them on the map.

**Reply by devemux86** [reply]
> Perhaps a small on-screen label (between the 2 buttons) like: "1/10", "2/10" etc to keep track of how far back you're going?

@MotoUKRider thanks, this an interesting idea.

The route planning history currently holds 10 changes.
Of course we can change it if needed.

**Comment by devemux86** [comment]
Undo / Redo in route planning is available in [Cruiser 3.0.23](https://play.google.com/store/apps/details?id=gr.talent.cruiser).

- **Android**: long press on map
- **Desktop**: right click on map

**Reply by DolomitesTom** [reply]
cool works

**Reply by devemux86** [reply]
Everyone, if you think of any improvements please let me know.
We started with a simple non annoying implementation. :slightly_smiling_face:

Also, I have hidden the route history screen.
Now with the undo + redo we don't need it?

**Comment by 0709wiwiwi** [comment]


**Reply by devemux86** [reply]
> map even zooms out completely

No, it just shows the whole route.
We can avoid this if we want, I wait feedback from the users.

Also I still wait for a complete workflow - if everyone wants fixed buttons - to describe how it can work.

Currently nobody offered a nice not annoying solution
and the apps that presented as examples were not user friendly, only had screen filled with buttons. :slightly_smiling_face:

**Reply by devemux86** [reply]
We all need to understand that we cannot fill the screen with buttons that appear all the time.

We already have 3+3 buttons appearing with each press on the map.
Do we need all of them? Can we use less buttons for more functions?

So undo / redo with 2 extra buttons? Appearing when? In route planning?
But route planning is always available. So currently this is not solved. :slightly_smiling_face:

**Reply by WolfgangZ42** [reply]
> We can avoid this if we want, I wait feedback from the users.

IMHO the map should not change after undo - if I need an overview I can select it by long press on the map. 
I agree with "no extra buttons" that just would save one tap.

**Reply by devemux86** [reply]
> IMHO the map should not change after undo - if I need an overview I can select it by long press on the map.

Map position and zoom with undo / redo will not change in the next Cruiser update.

We just probably need it (?) when the route after undo / redo is completely outside the visible map area.

**Reply by WolfgangZ42** [reply]
> when the route after undo / redo is completely outside the visible map area

That's a good point - so if the map doesn't show any route after undo it may switch to an overview of the whole route.

**Reply by devemux86** [reply]
In [Cruiser 3.0.25](https://play.google.com/store/apps/details?id=gr.talent.cruiser) the undo / redo actions do not auto center / zoom the map,
unless the route is outside the visible screen.

**Comment by 0709wiwiwi** [comment]
**Buttons fear?**  Compare by the gif image to find out a most optimal button **on map** location.

![Undo_Close_Redo_Buttons](https://github.com/devemux86/cruiser/assets/54167201/f0337022-442f-46f2-a957-3d651bae8a5d)

**Reply by devemux86** [reply]
> After export or navigation start, the undo/redo counter resets and thus buttons away.

Why export route or start navigation should clear route history? This will confuse users.
We make changes, export a route and then remember that we want the previous route.

In office applications, the undo is not cleared when we save documents.
After any change is made, then undo is available until we close the app.

**Reply by devemux86** [reply]
> I export a route when I think it is finished correctly.

It may be just a temporary export for backup, while route planning continues.
Each user uses the app functions differently. :slightly_smiling_face: 

> If I still want the original back later I re-import it.

I meant after any export, we want to get back to the route _before_ the export.
If export clears history, then this is not possible and the undo loses its value.

In any case automatic actions (like clear) in applications are not nice.

Apps must not have only 3-4 settings and force users to use them according to the developer's wishes.
Users should be masters of the apps and use them freely as they want.

Adding more buttons at the middle is not very practical,
as this space is occupied by the position / waypoint buttons.

Undo / redo small buttons can be added above the 3+3 existing buttons.
So we will have 4+4 buttons left + right (or at the bottom in landscape).

Always visible or auto-hide until restart the app and history is cleared.

**Reply by WolfgangZ42** [reply]
> clumsy undo function

I wouldn't call it clumsy and the "extra menu" presents the undo/redo buttons at once. As I said only an extra click ..,

<img width="500" src="https://github.com/devemux86/cruiser/assets/121806238/582e2a5b-fdf4-4676-a5f8-f977f55b9b66"/>

A click anywhere on the map 

<img width="500" src="https://github.com/devemux86/cruiser/assets/121806238/d38ecd65-4069-4bf1-b5fd-c54ea2ab1360"/>

A further click for undo/redo 

I guess we will not use undo/redo very often - why spend extra buttons for it? Regardless of my opinion - I'll live with any solution ...

**Comment by DolomitesTom** [comment]
> Also, I have hidden the route history screen.
Now with the undo + redo we don't need it?

have to think about it, I didn't found the history too bad

**Comment by 0709wiwiwi** [comment]
![Undo_Close_Redo_Buttons](https://github.com/devemux86/cruiser/assets/54167201/28ede67d-b0c1-4d51-9098-abd78688ae06)

Undo - Close - Redo buttons discreetly in the same position as the navigation simulation buttons.
Once a design is to your liking press the button X and gone are those then redundant buttons.
Simple & Discrete & Effective.

**Reply by 0709wiwiwi** [reply]
Experience Cruiser v 3.028.

The Undo ]X] Redo ON MAP button is becoming increasingly necessary.  A real route saver!
With more possibilities, the chance of making a mistake increases. That undo/redo button will one day be one of the most appreciated function. With the [X] you quickly hide the undo_redo buttons from the map display.

**Reply by devemux86** [reply]
Current undo / redo implementation works properly without being annoying and cluttering the screen.
(functionality does not change, only the presentation)

I mentioned that I might consider an implementation with on-screen buttons.
Such UI changes need careful consideration as the workflow is more complex in action.

**Reply by WolfgangZ42** [reply]
After pressing (X) the undo/redo buttons disappear - okay. But how do they come back? 

- Just by tap on the map, like in todays solution? Then you have to remember that "hidden feature" as well. 
- Or after each modifikation? Then you have to tap on (X) very often.

**Reply by devemux86** [reply]
Show undo / redo buttons all the time, even if they auto hide, is not something everyone likes.
Remember we should be able to use them at any time, even in navigation to undo a rerouting.

> After pressing (X) the undo/redo buttons disappear

Not just disappear, the X will close them, we will have to activate them again somehow. See steps below.

> - Just by tap on the map, like in todays solution? Then you have to remember that "hidden feature" as well.

And this requires more steps:
1. Press e.g. map to show undo / redo activation
2. Press undo / redo activation to display the buttons
3. Press undo or redo button
4. Press X to close them again

**Reply by 0709wiwiwi** [reply]
**Hiding is not closing the history**.  
As soon you trigger a subsequent route change action, the buttons are back there automatically. 
When the job is **completely** finished than you will tap the [X].  Anyway **all** buttons auto-hides.

**Reply by devemux86** [reply]
@0709wiwiwi do you mean the X hides the undo / redo buttons and clears the routing history,
until we do the next route change when they reappear automatically?

What about navigation? After a rerouting we want the previous route.
How this can be done without extra buttons in navigation UI?

**Reply by WolfgangZ42** [reply]
> No? 

**NO!!!** Sorry, but I don't need to see undo/redo buttons at any change of the route - for only let say 5 % undos I expect ...

**Reply by devemux86** [reply]
Users know from all programs and expect that undo / redo are available in all changes of the related element.

Please everyone see how editors work, we are not trying to reinvent the wheel here.
Simple things and simple implementation.

**Reply by devemux86** [reply]
By the way we are not talking about changes in how the undo / redo currently work.
It is extremely useful to have them (like routing history in the past) always available.

I repeat that it is still not clear how the extra buttons can help the existing implementation.
Considering that the functions remain the same, only the appearance changes.

**Reply by 0709wiwiwi** [reply]
Again technically the undo/redo works as expected and is 100% fine and usefull.
Only in terms of the tool presentation Wolfgang has a different view than I do.
As Cruiser is beginning to be as complicated to operate as the rivals.

**Reply by devemux86** [reply]
> As Cruiser is beginning to be as complicated to operate as the rivals

Adding more buttons probably makes the user interface more complicated. :slightly_smiling_face:

**Reply by devemux86** [reply]
@0709wiwiwi there is no problem, your and everyone's ideas are always useful! :slightly_smiling_face: 

It is understandable that users have features that they like more than others.
Settings have reasonable defaults, after some time with the app, users can start exploring them.

Cruiser platform has 1 developer all these years.
We need to maintain a clear design for the new features so that they can be implemented.

**Reply by devemux86** [reply]
@0709wiwiwi 

>  If you want to go back several (3) steps it is less practical.
long click map>click button>see map # long click map>click button>see map # long click map>click button>see map.

Agreed, quick undo takes more steps for now, this can be improved.

We can certainly improve the undo / redo workflow, let's see how... :slightly_smiling_face:

**Reply by Cer-o** [reply]
Estoy de acuerdo en que complicar en exceso la aplicación la hará menos atractiva. Espero que esto se tenga en cuenta durante su desarrollo.

**Comment by DolomitesTom** [comment]
Not every feature is immediately accessible. But once you have used, understood and got used to the new and old powerful features of Cruiser, you won't want to miss them. Every other navigation tool seems bland in comparison. How can you ever get away from it, it's addictive 🙄 😎 😁...

**Reply by devemux86** [reply]
Important is that all features can always be improved based on your suggestions.
That's why we always need your ideas so we can discuss possible improvements.

Also my suggestion remains: _try to break Cruiser_ or use it in unusual ways!
This is a good way to make it even more reliable for your outdoor activities.

**Comment by Ciutnik** [comment]
My trivial suggestion is to add the word "Undo" to the arrow, because at first glance it is easy to mistake it for a some "Back". 
If there is not enough space (especially when "Redo" also appears), maybe replace the arrows with words at all? I am thinking here about people starting their adventure with Cruiser.

**Comment by MotoUKRider** [comment]
I agree with the earlier comments about it (Undo) being an almost hidden feature. I had to read the announcement post further up to remember exactly how it worked (and I'd already used+contributed to this thread!), so I think it needs making more obvious but in a "don't get in my way too much" regard.

It *might* work on the Waypoints dialog (using those rounded arrows placed to the left of "Cancel/OK") where you can easily amend edits.  Again, potential danger of clutter. Tricky!

**Reply by 0709wiwiwi** [reply]
@MotoUKRider 
Compare this functionality to a house butler service.
_Provides spontaneous services when useful and discreetly withdraws after a simple gesture._

**Comment by devemux86** [comment]
Undo / Redo buttons will be available in next Cruiser.

- Visible in route planning (with auto hide)
- Press "X" to remove the buttons
- Undo / redo in long press is an alternative (when buttons are closed)

<img src="https://github.com/devemux86/cruiser/assets/3484020/2fcc7417-a039-43be-b974-aa60d51baa52" width="50%"/>

**Reply by devemux86** [reply, marked as answer]
Undo / Redo buttons are available in [Cruiser 3.0.30](https://play.google.com/store/apps/details?id=gr.talent.cruiser).

**Reply by WolfgangZ42** [reply]
Finaly I'm getting used to the undo/redo buttons and have to admit that they don't bother me and are even better than the previous approach: 
<img width="500" src="https://github.com/devemux86/cruiser/assets/121806238/923b28b6-e555-4fbf-9b89-3b7c13980b61"/> 
👍🥴

**Reply by devemux86** [reply]
Remember that if you close with the "X" the undo / redo buttons,
they remain available in long press menu, as an alternative way.
